### PR TITLE
AP_GPS: faster GPS lock

### PR DESCRIPTION
**Pull request by peterbarker**
use 9600 and 115200 first

These are the most common baudrates used for communicating with the GPS.  Probe them first and twice before trying the uncommon baud rates.

### CODE REVIEW

**Comment by WickedShell** [comment]
You want to keep 38400 infront of 115200 or all old 3DR GPS's that were sent out preconfigured will fall back in detection time (because we have to send the config string then loop till 115200 again). Dropping 19200 back makes sense, but I'm not sure if we want to drop 38400 yet.

**Comment by peterbarker** [comment]
@WickedShell don't we reconfigure them in a permanent fashion the first time we see them?

**Comment by WickedShell** [comment]
@peterbarker it's an opt-in behavior GPS_SAVE_CFG defaults to 0. Changing it to be a default of 2 would be a sane change though I think.

File: libraries/AP_GPS/AP_GPS.cpp [hunk 1]
```diff
@@ -55,7 +55,7 @@
 extern const AP_HAL::HAL &hal;
 
 // baudrates to try to detect GPSes with
-const uint32_t AP_GPS::_baudrates[] = {4800U, 19200U, 38400U, 115200U, 57600U, 9600U, 230400U};
+const uint32_t AP_GPS::_baudrates[] = {9600U, 115200U, 9600U, 115200U, 4800U, 19200U, 38400U, 57600U, 230400U};
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS.cpp]: Why is 9600U here twice?

Comment (OXINARF) [reply libraries/AP_GPS/AP_GPS.cpp]: @peterbarker wrote in the PR description that it now probes twice 9600 and 115200. That doesn't seem good to me; why is that needed?

**Comment by OXINARF** [comment]
By the way, the commit message (and PR title) is very misleading. I thought this was going to give a faster 3D lock. But after all this might just help get the driver loaded quicker.

File: libraries/AP_GPS/AP_GPS.cpp (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_GPS/AP_GPS.cpp]: In case the GPS isn't ready the first time we ask it at the most probable baudrates.

Comment (WickedShell) [reply libraries/AP_GPS/AP_GPS.cpp]: Do you have any log that shows this ever actually occuring? The GPS should have always beaten us on boot time and being ready to listening to the serial port. Doubling up here just slows down other baud rates/changes so I'd prefer to not do it on spec.

**Comment by peterbarker** [comment]
@OXINARF  GPS lock for the vehicle :-)

Also, depending on the configuration we shove in after we find the GPS, it could possibly be a faster lock for the GPS.  e.g. on the SkyViper we send in the mga-offline data (and various other bits and pieces) which definitely helps it lock on faster.

**Comment by peterbarker** [comment]
@WickedShell Default changes to 2.  You didn't need to ask twice :-)

**Comment by WickedShell** [comment]
@peterbarker lol figured you'd be down with that one :)

**Comment by peterbarker** [comment]
On Mon, 27 Nov 2017, WickedShell wrote:

> Do you have any log that shows this ever actually occuring? The GPS should

Nope.

> have always beaten us on boot time and being ready to listening to the
> serial port. Doubling up here just slows down other baud rates/changes so
> I'd prefer to not do it on spec.

Removed the duplicates.

**Pull request merged**